gateleen-routing/src/test/java/org/swisspush/gateleen/routing/ForwarderTest.java

```diff
@@ -0,0 +1,132 @@
+      HeaderFunction headerFunction = HeaderFunctions.parseFromJson(headers);
+      rule.setHeaderFunction(headerFunction);
+    }
+    rule.setUrlPattern(extractedRule.getString("url"));
+    rule.setHost(DEFAULTHOST);
+    rule.setPort(DEFAULTPORT);
+    return rule;
+  }
+
+  @Before
+  public void setUp() {
+    vertx = Mockito.mock(Vertx.class);
+    loggingResourceManager = Mockito.mock(LoggingResourceManager.class);
+    monitoringHandler = Mockito.mock(MonitoringHandler.class);
+    httpClient = Mockito.mock(HttpClient.class);
+    storage = new MockResourceStorage(ImmutableMap.of(RULES_PATH, RULES));
+    logger = LoggerFactory.getLogger(ForwarderTest.class.getName());
+  }
+
+  @Test
+  public void testHeaderFunctionsWithHostHeader() {
+    Rule rule = extractRule("/ruleWithHostHeader");
+    Forwarder forwarder = new Forwarder(vertx, httpClient, rule, storage, loggingResourceManager,
+        monitoringHandler, USER_PROFILE_PATH);
+    MultiMap reqHeaders = new VertxHttpHeaders();
+    reqHeaders.add(HOST_HEADER, "oldHost1234");
+    String errorMessage = forwarder.applyHeaderFunctions(logger, reqHeaders);
+    Assert.assertNull(errorMessage);
+    Assert.assertEquals(reqHeaders.get(HOST_HEADER),"newHost");
+  }
+
+  @Test
+  public void testHeaderFunctionsDefaultWithHostHeader() {
+    Rule rule = extractRule("/ruleWithHostHeader");
+    Forwarder forwarder = new Forwarder(vertx, httpClient, rule, storage, loggingResourceManager,
+        monitoringHandler, USER_PROFILE_PATH);
+    MultiMap reqHeaders = new VertxHttpHeaders();
+    reqHeaders.add(HOST_HEADER, DEFAULTHOST + ":" + DEFAULTPORT);
+    String errorMessage = forwarder.applyHeaderFunctions(logger, reqHeaders);
+    Assert.assertNull(errorMessage);
+    Assert.assertEquals(reqHeaders.get(HOST_HEADER),"newHost");
+  }
+
+  @Test
+  public void testHeaderFunctionsWithoutHostHeader() {
+    Rule rule = extractRule("/ruleWithoutHeader");
+    Forwarder forwarder = new Forwarder(vertx, httpClient, rule, storage, loggingResourceManager,
+        monitoringHandler, USER_PROFILE_PATH);
+    MultiMap reqHeaders = new VertxHttpHeaders();
+    reqHeaders.add(HOST_HEADER, "oldHost:1234");
+    String errorMessage = forwarder.applyHeaderFunctions(logger, reqHeaders);
+    Assert.assertNull(errorMessage);
+    Assert.assertEquals(reqHeaders.get(HOST_HEADER),DEFAULTHOST + ":" + DEFAULTPORT);
+  }
+
+  @Test
+  public void testHeaderFunctionsDefaultWithoutHostHeader() {
+    Rule rule = extractRule("/ruleWithoutHeader");
+    Forwarder forwarder = new Forwarder(vertx, httpClient, rule, storage, loggingResourceManager,
+        monitoringHandler, USER_PROFILE_PATH);
+    MultiMap reqHeaders = new VertxHttpHeaders();
+    reqHeaders.add(HOST_HEADER, DEFAULTHOST + ":" + DEFAULTPORT);
+    String errorMessage = forwarder.applyHeaderFunctions(logger, reqHeaders);
+    Assert.assertNull(errorMessage);
+    Assert.assertEquals(reqHeaders.get(HOST_HEADER),DEFAULTHOST + ":" + DEFAULTPORT);
+  }
+
+
```

Comment:
**WARN**: I think we should add the following test to the mix:

```java
@Test
public void hostIsAcceptedIfBothOfInputAndHeaderFunctionsWishToUseTheSameHost() {
    Rule rule = extractRule("/ruleWithHostHeader");
    Forwarder forwarder = new Forwarder(vertx, httpClient, rule, storage, loggingResourceManager,
            monitoringHandler, USER_PROFILE_PATH);
    MultiMap reqHeaders = new VertxHttpHeaders();
    reqHeaders.add(HOST_HEADER, "newHost");
    String errorMessage = forwarder.applyHeaderFunctions(logger, reqHeaders);
    Assert.assertNull(errorMessage);
    Assert.assertEquals("newHost", reqHeaders.get(HOST_HEADER));
}
```

Reply:
This is a good catch. But it is a wrong assumption. We do not care anymore the header coming in from somewhere. Therefore the case where the incoming header matches a configured header doesn't make sense as it must be set to the default in this case.   Or with other words, the incoming Host header for this method could only be the default, anything else must result in the default and configuring in the rule a host header which is the same as the default doesn't make sense really as it again results in the default.